src/block/Composter.php

```diff
@@ -0,0 +1,190 @@
+		foreach (Facing::HORIZONTAL as $side) {
+			$boxes[] = $this->getSideCollisionBox($side);
+		}
+		return $boxes;
+	}
+
+	protected function getSideCollisionBox(int $face = Facing::NORTH) : AxisAlignedBB{
+		$empty = abs(15 - 2 * $this->composter_fill_level) - (int) ($this->composter_fill_level === 0);
+		return ($face === Facing::DOWN || $face === Facing::UP) ? AxisAlignedBB::one()->contract(2 / 16, 0, 2 / 16)->trim(Facing::UP, $empty / 16) : AxisAlignedBB::one()->trim(Facing::opposite($face), 14 / 16);
+	}
+
+	public function isEmpty() : bool{
+		return $this->composter_fill_level === 0;
+	}
+
+	public function isReady() : bool{
+		return $this->composter_fill_level === 8;
+	}
+
+	public function getComposterFillLevel() : int{
+		return $this->composter_fill_level;
+	}
+
+	/**
+	 * @throws \Exception
+	 */
+	public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
+		if ($player instanceof Player && !$player->isSneaking()) {
+			if ($this->compost($player, clone $item)) {
+				$item->pop();
+			}
+		}
+		return true;
+	}
+
+	public function pushCollidedEntities() : void{
+		if ($this->composter_fill_level === 7) return;
+		foreach (
+			$this->position->getWorld()->getNearbyEntities(
+				$this->getSideCollisionBox(Facing::DOWN)->extend(Facing::UP, 0.25)->offset(
+					$this->position->getFloorX(),
+					$this->position->getFloorY(),
+					$this->position->getFloorZ()
+				)
+			) as $entity) {
+			if ($entity instanceof Player || $entity instanceof Projectile) continue;
+			$motion = $entity->getMotion();
+			$motion->y = 0.2; //Lower can make entity clip through the block, the game cannot catch up the speed of the entity? :/
+			$entity->setMotion($motion);
+		}
+	}
+
+	protected function spawnParticleEffect() : void{
+		$packet = SpawnParticleEffectPacket::create(
+			DimensionIds::OVERWORLD,
+			-1,
+			$this->position->add(0.5, 0.5, 0.5),
+			self::CROP_GROWTH_EMITTER_PARTICLE
+		);
+		foreach ($this->position->getWorld()->getViewersForPosition($this->position) as $player) {
+			$player->getNetworkSession()->sendDataPacket($packet);
+		}
+	}
+
+	/**
+	 * @throws \Exception
+	 */
+	public function compost(Block|Player $origin, ?Item $item = null) : bool{
+		if ($this->composter_fill_level >= 8) {
+			$this->position->getWorld()->dropItem(
+				$this->position->add(0.5, 0.85, 0.5),
+				new Fertilizer(new ItemIdentifier(ItemIds::DYE, 15), "Bone Meal"),
+				new Vector3(0, 0, 0)
+			);
+			$this->position->getWorld()->addSound($this->position, new ComposterEmptySound());
+			$this->spawnParticleEffect();
+			$this->composter_fill_level = 0;
+			$this->position->getWorld()->setBlock($this->position, $this);
+			return false;
+		} else {
+			if ($item === null || !ComposterUtils::isCompostable($item)) return false;
+			$this->spawnParticleEffect();
+
+			$percent = ComposterUtils::getPercentage($item);
+			if (mt_rand(1, 100) <= $percent) {
+				$this->pushCollidedEntities();
```

Comment:
This is not something that should be done at a block-specific level (it's generally not implemented yet).

Reply:
so i should remove it

Reply:
you mean that `pushCollidedEntities `

Reply:
Yes